Disallow creation of ProcessGroupNCCL without GPUs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45642 Disallow creation of ProcessGroupNCCL without GPUs.**

Prior to https://github.com/pytorch/pytorch/pull/45181, initializing a
NCCL process group would work even if no GPUs were present. Although, now since
init_process_group calls `barrier()` this would fail.

In general the problem was that we could initialize ProcessGroupNCCL without
GPUs and then if we called a method like `barrier()` the process would crash
since we do % numGPUs resulting in division by zero.

Differential Revision: [D24038839](https://our.internmc.facebook.com/intern/diff/D24038839/)

pritamdamania committed:
Disallow creation of ProcessGroupNCCL without GPUs.

Prior to https://github.com/pytorch/pytorch/pull/45181, initializing a
NCCL process group would work even if no GPUs were present. Although, now since
init_process_group calls `barrier()` this would fail.

In general the problem was that we could initialize ProcessGroupNCCL without
GPUs and then if we called a method like `barrier()` the process would crash
since we do % numGPUs resulting in division by zero.

Differential Revision: [D24038839](https://our.internmc.facebook.com/intern/diff/D24038839/)

[ghstack-poisoned]

osalpekar reviewed (approved):
Will this be cherry-picked into 1.7, since  #45181 made it in?

pritamdamania87 commented:
> Will this be cherry-picked into 1.7, since #45181 made it in?

Yes, that is the plan.

pritamdamania committed:
Update on "Disallow creation of ProcessGroupNCCL without GPUs."


Prior to https://github.com/pytorch/pytorch/pull/45181, initializing a
NCCL process group would work even if no GPUs were present. Although, now since
init_process_group calls `barrier()` this would fail.

In general the problem was that we could initialize ProcessGroupNCCL without
GPUs and then if we called a method like `barrier()` the process would crash
since we do % numGPUs resulting in division by zero.

Differential Revision: [D24038839](https://our.internmc.facebook.com/intern/diff/D24038839/)

[ghstack-poisoned]
